src/core_plugins/kibana/server/lib/management/saved_objects/relationships.js

```diff
@@ -0,0 +1,222 @@
+    const visualizationResponse = await callCluster('mget', {
+      body: {
+        docs: visualizationIds.slice(0, size).map(id => ({
+          _index: kibanaIndex,
+          _type: 'doc',
+          _id: `visualization:${id}`,
+          _source: [`visualization.title`]
+        }))
+      }
+    });
+
+    visualizations.push(...visualizationResponse.docs.reduce((accum, doc) => {
+      if (doc.found) {
+        accum.push({
+          id: formatId(doc._id),
+          title: doc._source.visualization.title,
+        });
+      }
+      return accum;
+    }, []));
+  }
+
+  return { visualizations };
+}
+
+async function findVisualizationRelationships(id, size, callCluster, savedObjectsClient) {
+  const kibanaIndex = savedObjectsClient._index;
+  const allDashboardsResponse = await callCluster('search', {
+    index: kibanaIndex,
+    size: 10000,
+    ignore: [404],
+    _source: [`dashboard.title`, `dashboard.panelsJSON`],
+    body: {
+      query: {
+        term: {
+          type: 'dashboard'
+        }
+      }
+    }
+  });
+
+  const dashboards = [];
+  for (const dashboard of allDashboardsResponse.hits.hits) {
+    const panelsJSON = JSON.parse(dashboard._source.dashboard.panelsJSON);
+    if (panelsJSON) {
+      for (const panel of panelsJSON) {
+        if (panel.type === 'visualization' && panel.id === id) {
+          dashboards.push({
+            id: formatId(dashboard._id),
+            title: dashboard._source.dashboard.title,
+          });
+        }
+      }
+    }
+
+    if (dashboards.length >= size) {
+      break;
+    }
+  }
+
+  return { dashboards };
+}
+
+async function findSavedSearchRelationships(id, size, callCluster, savedObjectsClient) {
+  const kibanaIndex = savedObjectsClient._index;
+  const search = await savedObjectsClient.get('search', id);
+
+  const searchSourceJSON = JSON.parse(search.attributes.kibanaSavedObjectMeta.searchSourceJSON);
+
+  let indexPatterns = [];
+  try {
+    const indexPattern = await savedObjectsClient.get('index-pattern', searchSourceJSON.index);
+    indexPatterns = [{ id: indexPattern.id, title: indexPattern.attributes.title }];
+  } catch (err) {
+    // Do nothing
+  }
+
+  const allVisualizationsResponse = await callCluster('search', {
+    index: kibanaIndex,
+    size,
+    ignore: [404],
+    _source: [`visualization.title`],
+    body: {
+      query: {
+        term: {
+          'visualization.savedSearchId': id,
+        }
+      }
+    }
+  });
+
+  const visualizations = allVisualizationsResponse.hits.hits.map(response => ({
+    id: formatId(response._id),
+    title: response._source.visualization.title,
+  }));
+
+  return { visualizations, indexPatterns };
+}
+
+async function findIndexPatternRelationships(id, size, callCluster, savedObjectsClient) {
+  const kibanaIndex = savedObjectsClient._index;
+
+  const [allVisualizationsResponse, savedSearchResponse] = await Promise.all([
+    callCluster('search', {
```

Comment:
You could use an `_msearch` here to send both at the same time as a single request. I kind of wonder if it may be worth fetching both together?

```js
query: {
  bool: {
    filter: [
      { exists: { field: 'visualization.kibanaSavedObjectMeta.searchSourceJSON' } },
      {
        bool: {
          should: [
            { term: { type: 'visualization' } },
            { term: { type: 'search' } }
          ]
        }
      }
    ]
  }
},
_source: ['*.title', '*.kibanaSavedObjectMeta.searchSourceJSON'],
size: 10000,
sort: 'type'
```

Sorting by type effectively groups the types together. This does however increase the unlikely opportunity for > 10000 saved objects causing issues (given this method's purpose).